### PR TITLE
proxy-protocol-filter: add version to filter state

### PR DESCRIPTION
**Pull request by fl0Lec**
Commit Message: proxy-protocol-filter: add version to filter state
Additional Description: allows to differentiate filter proxy-protocol filter state create by proxy-protocol filter with the one created by HCM. 
Risk Level: low
Testing: true
Docs Changes: no
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

Context
--------
Currently the proxy-protocol-filter will set a filter state metadata with the following field:
- source address
- destination address
- tlv_vector

By default if at the Http Connection Manager (HCM) they are no filter state for proxy-protocol, it will set one with the source and destination address.

 Currently they are no way to know from that filter state if the tlv_vector is empty because there was no entry or either because proxy-protocol v1 was used or because the HCM created the state.

This is making the filter state more clear, where the version:
- **NotUsed**: the state was set by another filter
- **NotFound**: proxy-protocol filter was present but did not found any proxy-protocol in the TCP connection
- **V1**: successfully parse proxy-protocol filter V1, TLV vector is empty
- **V2**: successfully parse proxy-protocol filter V1, TLV vector will be populated with entries found

### CODE REVIEW

File: source/common/http/conn_manager_impl.cc [hunk 1]
```diff
@@ -171,7 +172,8 @@ void ConnectionManagerImpl::initializeReadFilterCallbacks(Network::ReadFilterCal
         Network::ProxyProtocolFilterState::key(),
         std::make_unique<Network::ProxyProtocolFilterState>(Network::ProxyProtocolData{
             read_callbacks_->connection().connectionInfoProvider().remoteAddress(),
-            read_callbacks_->connection().connectionInfoProvider().localAddress()}),
+            read_callbacks_->connection().connectionInfoProvider().localAddress(),
+            Envoy::Network::ProxyProtocolVersion::NotFound}),
```
[the remark this answers is not in the document]
Comment (fl0Lec) [reply source/common/http/conn_manager_impl.cc]: Yes sure, I added a `NotUsed` proxy-protocol version with some comment to keep. Let me know if that works for you

**Review by ggreenway — CHANGES_REQUESTED** [review]
This version doesn't build, with error:

`source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:185:11: error: enumeration value 'NotUsed' not handled in switch`

That helps point out what's weird about a value that's only valid in some contexts.

Instead of adding a new value to the enum, could you add a new struct that inherits from `ProxyProtocolData` and adds the enum, and then use the appropriate type in the right contexts (inputs vs outputs)?

/wait

**Comment by fl0Lec** [comment]
I had misunderstood your comment about input/output in the previous comment, if my understanding is correct:
the proxy-protocol data filter state is set:
- by the proxy-protocol filter if present
- by either the http connection manager or tcp proxy filter otherwise
That's our input state.

This allows to forward the source/destin ip/port in the case where proxy-protocol is used upstream however the version field would be irrelevant here.

I propose to have a `ProxyProtocolDataWithVersion` which is a derived struct of `ProxyProtocolData`. This will allows to avoid any confusion for the function that want to use `ProxyProtocolData` while still allows other filter to get the version of the proxy-protocol that was used.

Does that makes sense for you ?

**Comment by ggreenway** [comment]
Yep, I think that makes sense, and should make the code clearer.

**Comment by fl0Lec** [comment]
> Yep, I think that makes sense, and should make the code clearer.

I already pushed the changed and it is ready for review :)

File: envoy/network/proxy_protocol.h [hunk 2]
```diff
@@ -26,5 +26,18 @@ struct ProxyProtocolData {
...
+  // the proxy-protocol version.
+  NotUsed = 0,
+  // The proxy-protocol filter did not find any proxy-protocol header and was skipped.
+  NotFound = 1,
+  V1 = 2,
+  V2 = 3
+};
+
+struct ProxyProtocolDataWithVersion : public ProxyProtocolData {
+  const ProxyProtocolVersion version_;
```
Comment (ggreenway) [inline envoy/network/proxy_protocol.h:40]: You can avoid needing a separate enum with `NotUsed` by making this an `absl::optional`. Then you don't need to cast between enum types.

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc [hunk 3]
```diff
@@ -222,6 +222,18 @@ ReadOrParseState Filter::parseBuffer(Network::ListenerFilterBuffer& buffer) {
       return read_header_state;
     }
     if (header_version_ == ProxyProtocolVersion::NotFound) {
+      if (!cb_->filterState().hasData<Network::ProxyProtocolFilterState>(
```
Comment (ggreenway) [inline source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:225]: This block seems incorrect. You're setting the proxy protocol filter state when there isn't any proxy protocol. `NotFound` seems implied by the absence of this filter state.

**Review by ggreenway — CHANGES_REQUESTED** [review]
/wait

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc (hunk 3 above)
Comment (fl0Lec) [reply source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc]: I would have like t have a ways to differentiate between proxy-protocol filter was present but did not found any proxy-protocol in the connection and they are no proxy-protocol filter. But I am fine with removing it.

**Comment by fl0Lec** [comment]
Thanks for the review, I appreciate it

**Review by ggreenway — APPROVED** [review]

**Comment by fl0Lec** [comment]
Thank you for the review @/ggreenway

**Pull request merged**